snapcraft/parts/setup_assets.py

```diff
@@ -0,0 +1,215 @@
+) -> Optional[Path]:
+    """Ensure sure icon is properly configured and installed.
+
+    Fetch from a remote URL, if required, and place in the meta/gui
+    directory.
+    """
+    # Nothing to do if no icon is configured, search for existing icon.
+    if icon is None:
+        return _find_icon_file(assets_dir)
+
+    # Extracted appstream icon paths will either:
+    # (1) point to a file relative to prime
+    # (2) point to a remote http(s) url
+    #
+    # The 'icon' specified in the snapcraft.yaml has the same
+    # constraint as (2) and would have already been validated
+    # as existing by the schema.  So we can treat it the same
+    # at this point, regardless of the source of the icon.
+    parsed_url = urllib.parse.urlparse(icon)
+    parsed_path = Path(parsed_url.path)
+    icon_ext = parsed_path.suffix[1:]
+    target_icon_path = Path(gui_dir, f"icon.{icon_ext}")
+
+    target_icon_path.parent.mkdir(parents=True, exist_ok=True)
+    if parsed_url.scheme in ["http", "https"]:
+        # Remote - fetch URL and write to target.
+        emit.progress(f"Fetching icon from {icon!r}")
+        icon_data = requests.get(icon).content
+        target_icon_path.write_bytes(icon_data)
+    elif parsed_url.scheme == "":
+        source_path = Path(
+            prime_dir,
+            parsed_path.relative_to("/") if parsed_path.is_absolute() else parsed_path,
+        )
+        if source_path.exists():
+            # Local with path relative to prime.
+            shutil.copy(source_path, target_icon_path)
+        elif parsed_path.exists():
+            # Local with path relative to project.
+            shutil.copy(parsed_path, target_icon_path)
+        else:
+            # No icon found, fall back to searching for existing icon.
+            return _find_icon_file(assets_dir)
+    else:
+        raise RuntimeError(f"Unexpected icon path: {parsed_url!r}")
+
+    return target_icon_path
+
+
+def _find_icon_file(assets_dir: Path) -> Optional[Path]:
+    for icon_path in (assets_dir / "gui/icon.png", assets_dir / "gui/icon.svg"):
+        if icon_path.is_file():
+            return icon_path
+    return None
+
+
+def _validate_command_chain(
+    command_chain: List[str], *, app_name: str, prime_dir: Path
+) -> None:
+    """Verify if each item in the command chain is executble."""
+    for item in command_chain:
+        executable_path = prime_dir / item
+
+        # command-chain entries must always be relative to the root of
+        # the snap, i.e. PATH is not used.
+        if not _is_executable(executable_path):
+            raise errors.SnapcraftError(
+                f"Failed to generate snap metadata: The command-chain item {item!r} "
+                f"defined in the app {app_name!r} does not exist or is not executable.",
+                resolution=f"Ensure that {item!r} is relative to the prime directory.",
+            )
+
+
+def _is_executable(path: Path) -> bool:
```

Comment:
isn't this a bit too much? Also, we are checking for is_file here, not is_executable